### ISSUE TITLE
New specs for Jakarta EE 9

### ISSUE THREAD

**Issue by bjhargrave** [osgi/osgi#54]
Original bug ID: [BZ#3145](https://osgi.org/members/bugzilla/show_bug.cgi?id=3145)
From: @tjwatson
Reported version: R8

**Comment by cziegeler** [comment]
I assume we will handle this on a spec by spec base where we might handle it differently for each spec. We might also not update some specs like the http service.

Should we maybe create separate issues about for a spec we want to work on?

**Comment by bjhargrave** [comment]
> Should we maybe create separate issues about for a spec we want to work on?

Yes. This is an umbrella issue. When you create an issue for a specific spec, please make sure to update this issue to mention it.

**Comment by bjhargrave** [comment]
I will go through the Compendium specs and make a recommendation about which specs to update to jakarta and which specs will not be updated and removed from Compendium.

**Comment by cziegeler** [comment]
What is the general plan? Will new specs only support api from the jakarta namespace? Or both? Or do we decide per spec?

**Comment by timothyjward** [comment]
> What is the general plan? Will new specs only support api from the jakarta namespace? Or both? Or do we decide per spec?

If we can simplify the migration/coexistence story then that's best for users.

If Felix would have to package two copies of Jetty to support the old and new packages then that's clearly impractical, but if it's possible to support both apis from a single implementation bundle then I think we should.

**Comment by cziegeler** [comment]
I would assume that you can handle supporting the old packages via wrappers - should at least work for http whiteboard.
I would prefer that as well

**Comment by bjhargrave** [comment]
I think the plan would be that Compendium R9 would be jakarta only. Compendium R8 does not cease to exist, so people can implement and use R8 based impls. Impls of R9 can support old javax if they want but that would not be an R9 spec requirement and the R9 TCK would test jakarta only.

I don't want to have an R9 spec requiring support for both jakarta and javax. Doing so would be an impl choice and not tested by the spec's TCK.

**Comment by cziegeler** [comment]
ok, fair enough - in that case, we must make sure that implementations can support javax packages without a high burden. For example, if we have an api package in version 1.3 which uses a javax package as part of their signature, changing that to jakarta packages and updating the major version to 2.0 sounds reasonable from an api and also user perspective. But it makes it very hard if not impossible for an implementation to implement both.

**Comment by bjhargrave** [comment]
Well the obvious choice is to bump the spec major version and the spec API package major version and replace javax API references with jakarta API references. Renaming the spec API packages (so that the javax referring package names and the jakarta referring package names are distinct) seems a large burden on everyone for the very small community of implementors who want to implement both old and new spec in the same impl bundle (for which there is little evidence this is practical or even possible).

**Comment by tjwatson** [comment]
> ok, fair enough - in that case, we must make sure that implementations can support javax packages without a high burden. For example, if we have an api package in version 1.3 which uses a javax package as part of their signature, changing that to jakarta packages and updating the major version to 2.0 sounds reasonable from an api and also user perspective. But it makes it very hard if not impossible for an implementation to implement both.

To use the HttpService as an example.  I think this is a case where we would deprecate (or actually remove it from the specification).  An HTTP Whiteboard implementation could choose to still understand javax Servlet registrations but would not be required to.  But we do have a difficult time with org.osgi.service.http.context.ServletContextHelper.

**Comment by cziegeler** [comment]
Right, I'm exactly worried about that class - if we provide a 2.0 API with a ServletContextHelper bound to the jakarata package

**Comment by timothyjward** [comment]
> Well the obvious choice is to bump the spec major version and the spec API package major version and replace javax API references with jakarta API references. Renaming the spec API packages (so that the javax referring package names and the jakarta referring package names are distinct) seems a large burden on everyone for the very small community of implementors who want to implement both old and new spec in the same impl bundle (for which there is little evidence this is practical or even possible).

A burden also applies to users of the specification if only the jakarta API namespace is supported. If, for example, I have an existing runtime and I want to add a new JAX-RS application or a new Servlet then I either:

* Have to keep using the old javax API namespace in my new code
* Upgrade all my existing code to the new Jakarta namespace
* Run more than one whiteboard implementation (one for javax and one for jakarta) on different ports

All of these options are suboptimal...

It's also the case that a lot of existing libraries and tools provide types using the javax APIs, are we saying that from R9 those will be unavailable?

**Comment by bjhargrave** [comment]
I am not sure what you are arguing for? Doing nothing about jakarta? I don't think that it practical.

We either have to bump the major version of the API packages and people can use the current R8 APIs/impls or use the new R9 APIs/impls. Or we create entirely new package names (`org.osgi.service.jaxrs.client` becomes ??) for all the API packages for a very small set of use cases.

I am not concerned by the use case of updating an existing bundle which uses javax-referring API to add a new jakarta-referring API part. You will need to update the entire bundle.

Supporting both javax-referring and jakarta-referring whiteboards can be done. If they can share the same port is more of an impl question for the whiteboard impls. Such dual-faced impls can certainly be structured (through a set of bundles) to do this.

I do not want to OSGi to have dual-faced specs which simultaneously support  javax-referring and jakarta-referring APIs. That will lead to madness. We, like others in the industry, will need to make a cutover. Implementations are free to support both faces, Open Liberty does this quite well and the Jakarta EE specs do not require them to. But each application on Open Liberty must be either javax-referring of jakarta-referring. There is no reasonable space where an application can be a mis-mash of both. Pick a side! :-)

**Comment by cziegeler** [comment]
Picking a side is not really user friendly - I understand thats what servlet engine implementations are currently doing, but we can do better. 
Regardless, at least for http whiteboard, the spec I'm concerned most, it will work with a 2.0 version of http whiteboard using jakarata packages - that's because there is no provided service defined in the spec, just  whiteboard interfaces. The runtime service is agnostic of the package space.. An implementation can then support R7 and R9 in parallel

**Comment by bjhargrave** [comment]
> Picking a side is not really user friendly

The whole jakarta namespace change is not user friendly... yet here we are :-)

**Comment by timothyjward** [comment]
> The runtime service is agnostic of the package space

Except for the `ServletContextHelper`, which references `HttpServletRequest` and `HttpServletResponse`

The JAX-RS whiteboard may be safe, but I'm not yet certain. Transaction Control may be able to get away with only making the JPA package version 2.0.

**Comment by tjwatson** [comment]
> > The runtime service is agnostic of the package space
> 
> Except for the `ServletContextHelper`, which references `HttpServletRequest` and `HttpServletResponse`
> 

If we want to leave the option open for implementations we could pick a new package name for the types that reference jakarta types.  For example, maybe `org.osgi.service.http.jakarta.context` makes sense for the HTTP Whiteboard Jakarta base spec.

**Comment by cziegeler** [comment]
> > The runtime service is agnostic of the package space
> 
> Except for the `ServletContextHelper`, which references `HttpServletRequest` and `HttpServletResponse`
> 
ah sorry, I meant everything in the http.runtime package - there is no need to have a 2.0 version for that
http.whiteboard and http.context have those reference - but supporting two different major versions shouldn't pose a problem

> The whole jakarta namespace change is not user friendly... yet here we are :-)
Touché - but just because someone does a mess to users, thats not a good reason to repeat that :)
I still believe for http.whiteboard we are fine - if the package for the HttpServiceRuntime service does not get a major version update.

**Comment by bjhargrave** [comment]
> For example, maybe `org.osgi.service.http.jakarta.context`

Egads, no! This is an inflection point, we do not want to bake it permanently into the api names.

A bundle must be either fully javax or fully jakarta. I do not want a goal of this work to be supporting bundles attempting to have parts be javax and parts be jakarta. Under this goal, there is no need to bake jakarta into package names. We either keep the package names and bump the major version or we completely rename all the packages to have a new parent package. I think the latter is confusing and wasteful of time arguing over naming :-) 

I think impls can support both old and new but not in one bundle. They will need multiple bundles to handle the package versions.

We must also be cognizant of the availability of our resources to address the jakarta issue. Bumping major package versions and refactoring to use jakarta API will be hard and time consuming enough. Attempting to support a dual-faced approach to the specs will be significantly harder and take much more time and effort to attempt.

**Comment by tjwatson** [comment]
> > For example, maybe `org.osgi.service.http.jakarta.context`
> 
> Egads, no! This is an inflection point, we do not want to bake it permanently into the api names.
> 
Fine, but we are already going to 2.0 for the spec version with breaking changes having a renamed package is perfectly acceptable to me.  I think we can come up with a name for the package to make it easier for impls that want to support both.  I'll just throw one more name out there `org.osgi.service.http.helpers`.  Arguably a better name to leave it open for adding other helpers and not make it so specific to context.

**Comment by cziegeler** [comment]
Should we move this discussion to a http whiteboard specific issue? As mentioned I see an easy path forward which hopefully will make BJ and everyone else happy :) I could create that one and outline the idea

**Comment by cziegeler** [comment]
Created #381 for http whiteboard

**Comment by bjhargrave** [comment]
The new Jakarta contracts are now available at https://docs.osgi.org/reference/portable-java-contracts.html.

**Comment by ideas-into-software** [comment]
Closely related: Upgrade Apache Felix to Jakarta Servlet API 6.x (https://issues.apache.org/jira/browse/FELIX-6612). 

There are quite a lot of prerequisites, along with questions which will further clarify next steps to take. Kindly please post your comments / clarifications via https://issues.apache.org/jira/browse/FELIX-6612 to move this forward.

**Comment by jluehe** [comment]
@bjhargrave, is there a Jakarta EE9 compatible version of `org.osgi:org.osgi.compendium` available? We are currently using version `5.0.0`. I see this artifact has been moved/renamed to `org.osgi:osgi.cmpn`, but its latest version: `7.0.0`, also still references `javax.servlet` instead of `jakarta`. Thank you for any pointers that you may have!

**Comment by bjhargrave** [comment]
See https://docs.osgi.org/artifacts/

> Other than the dedicated build time artifacts, it is generally recommended to use the individual artifacts, as this allows for more precise versioning. In particular, avoid using the osgi.cmpn artifact. After Release 7, the osgi.cmpn artifact is no longer published to Maven Central to encourage the use of the individual artifacts.

So do not use `osgi.cmpn` or `org.osgi.compendium` as they are no longer recommended for use nor published.

**Comment by jluehe** [comment]
> See https://docs.osgi.org/artifacts/
> 
> > Other than the dedicated build time artifacts, it is generally recommended to use the individual artifacts, as this allows for more precise versioning. In particular, avoid using the osgi.cmpn artifact. After Release 7, the osgi.cmpn artifact is no longer published to Maven Central to encourage the use of the individual artifacts.
> 
> So do not use `osgi.cmpn` or `org.osgi.compendium` as they are no longer recommended for use nor published.

Thank you, @bjhargrave, for this very helpful tip!
